lib/mayaUsd/fileio/instancedNodeWriter.cpp

```diff
@@ -168,7 +168,7 @@ UsdMaya_InstancedNodeWriter::UsdMaya_InstancedNodeWriter(
             // Replace prefixes to obtain DAG-USD path mapping.
             const UsdMayaUtil::MDagPathMap<SdfPath>& writerMapping
                 = writer->GetDagToUsdPathMapping();
-            for (const std::pair<MDagPath, SdfPath>& pair : writerMapping) {
```

Comment:
I understand the above cases, where we want to avoid copies (and it's such a common mistake)...but I'm curious what exactly is the compiler warning in this case?

Reply:
This was the compiler warning for that specific area.

```
/Users/dhruvgovil/Projects/maya-usd/lib/mayaUsd/fileio/instancedNodeWriter.cpp:171:54: warning: loop variable 'pair' has type 'const std::pair<MDagPath, SdfPath> &' but is initialized with type 'const std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<Autodesk::Maya::OpenMaya20200000::MDagPath, pxrInternal_v0_20__pxrReserved__::SdfPath>, std::__1::__tree_node<std::__1::__value_type<Autodesk::Maya::OpenMaya20200000::MDagPath, pxrInternal_v0_20__pxrReserved__::SdfPath>, void *> *, long> >::value_type' (aka 'const pair<const Autodesk::Maya::OpenMaya20200000::MDagPath, pxrInternal_v0_20__pxrReserved__::SdfPath>') resulting in a copy [-Wrange-loop-analysis]
            for (const std::pair<MDagPath, SdfPath>& pair : writerMapping) {
                                                     ^
/Users/dhruvgovil/Projects/maya-usd/lib/mayaUsd/fileio/instancedNodeWriter.cpp:171:18: note: use non-reference type 'std::pair<MDagPath, SdfPath>' to keep the copy or type 'const std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<Autodesk::Maya::OpenMaya20200000::MDagPath, pxrInternal_v0_20__pxrReserved__::SdfPath>, std::__1::__tree_node<std::__1::__value_type<Autodesk::Maya::OpenMaya20200000::MDagPath, pxrInternal_v0_20__pxrReserved__::SdfPath>, void *> *, long> >::value_type &' (aka 'const pair<const Autodesk::Maya::OpenMaya20200000::MDagPath, pxrInternal_v0_20__pxrReserved__::SdfPath> &') to prevent copying
            for (const std::pair<MDagPath, SdfPath>& pair : writerMapping) {
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Enabling `-Wrange-loop-analysis` suggested adding back the `&` so I put it back. I suspect it was a false positive of some kind without the explicit flag. Not exactly sure why it suggested removing it in the first place, because reading the actual code...it makes sense to me to leave the `&` in.